### PR TITLE
import and download spreadsheet for bulk updates

### LINKED ISSUE #13: Downloadable spreadsheet to stay up to date with any metadata updates that are made
Implementation - generate a spreadsheet with the data filled in as represented in Ingest.

Sub-tasks: (added 14/07/2021, issues found last sprint demo)
- [x] Fix caching and exported spreadsheet cleanup
- [x] Add the missing rows (the 1st 4 rows from the original spreadsheet with the column names and static text)
- [x] Add all the other data types (Karoly)
  - [x] Processes
  - [x] Protocols
  - [x] Data (files)
- [x] Fix tab names --- shouldn't be encounter module worksheets (Alegria)
- [x] integration / unit test (download & upload spreadsheet) (Karoly?) 
        https://github.com/ebi-ait/ingest-broker/pull/17/files
- [x] bug in spreadsheet generated - some header columns are missing like 'project.contributors.project_role.ontology_label' (added 15/07/2021) Alegria & Karoly
- [x] metadata row values are not aligned to correct column headers (added 16/07/2021) Alegria & Karoly
- [x] data rows should start at row 6 (16/07/2021) Karoly
- [x] project worksheet should go first before "project - contributors" etc (added 19/07/2021) - Karoly
- [x] submission errors are not being cleared (added 20/07/2021)
- [x] couldn't import spreadsheets with module worksheets like _collection protocol - reagents_ see 10x test spreadsheet flattener shouldn't add this as a separate module worksheet (added 20/07/2021)
- [x] run test plans with a wrangler
- [x] docs - wrangler will help. See [PR for SOP](https://github.com/ebi-ait/hca-ebi-wrangler-central/pull/386)

- [x] testing scenarios
   - [x] Enrique's scenarios - https://docs.google.com/spreadsheets/d/1hf6k7EUw5_l3mZEHVMi-iRZTRtCFyTDS1rOcIXNHBpo/edit#gid=0

### PR DESCRIPTION
**Pull request by amnonkhen**
endpoint & service for exporting project metadata to a spreadsheet
closes dcp-13
closes dcp-349

### CODE REVIEW

**Comment by aaclan-ebi** [comment]
The target branch should be dev. I think we could make dev the default branch in our repos to avoid forgetting this.

File: broker/service/spreadsheet_upload_service.py [hunk 1]
```diff
@@ -16,28 +18,55 @@ def __init__(self, ingest_api: IngestApi, storage_service: SpreadsheetStorageSer
...
         thread.start()
+
         return submission_resource
 
-    def _upload(self, submission_resource, path, is_update, project_uuid=None):
+    def _store_spreadsheet_updates(self, filename, request_file, submission_uuid):
+        submission_directory = self.storage_service.get_submission_dir(submission_uuid)
+        timestamp = time.strftime("%Y%m%d-%H%M%S")
+        filename_with_timestamp = f'{timestamp}_{filename}'
+        # TODO This spreadsheet containing the updates is not downloadable anywhere yet
```
Comment (ke4) [inline broker/service/spreadsheet_upload_service.py:44]: Do we need this comment here? If not, then better to remove it. Tx

File: broker/service/spreadsheet_upload_service.py [hunk 2]
```diff
@@ -16,28 +18,55 @@ def __init__(self, ingest_api: IngestApi, storage_service: SpreadsheetStorageSer
...
+            thread = threading.Thread(target=self.upload, args=(submission_url, path, project_uuid))
 
-        thread = threading.Thread(target=self._upload,
-                                  args=(submission_resource, path, is_update, project_uuid))
         thread.start()
+
         return submission_resource
 
-    def _upload(self, submission_resource, path, is_update, project_uuid=None):
+    def _store_spreadsheet_updates(self, filename, request_file, submission_uuid):
```
Comment (ke4) [inline broker/service/spreadsheet_upload_service.py:40]: Do we need to store the uploaded spreadsheets after we made work the import -> export -> modify -> import flow?

**Review by aaclan-ebi — DISMISSED** [review]
Approving for those changes i haven't made ;)

**Review by aaclan-ebi — DISMISSED** [review]

File: requirements.txt [hunk 3]
```diff
@@ -1,14 +1,11 @@
...
 et-xmlfile==1.0.1
 Flask-Cors==3.0.3
-idna==2.6
 itsdangerous==0.24
-jdcal==1.4
 MarkupSafe==1.1
 werkzeug>=0.15.3
--e git+https://github.com/ebi-ait/ingest-client.git@7f04ab9a#egg=hca_ingest
+-e git+https://github.com/ebi-ait/ingest-client.git@7038dcb#egg=hca_ingest
+#-e git+https://github.com/ebi-ait/ingest-client.git@6fd08d7#egg=hca_ingest
```
Comment (ke4) [inline requirements.txt]: Could we remove this commented out requirement, please?

**Review by ke4 — CHANGES_REQUESTED** [review]
Just a tiny request, but otherwise looks fine 👍

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -1,2 +1,13 @@
...
+  remote: 'https://raw.githubusercontent.com/ebi-ait/gitlab-ci-templates/master/build-release-deploy.yml'
+
+Unit Test:
+  image: quay.io/ebi-ait/ingest-base-images:python_3.6-slim
+  variables:
+    INGEST_API : https://api.ingest.dev.archive.data.humancellatlas.org/
+  script:
+    - apt-get update
+    - apt-get install -y git
+    - pip install -r requirements.txt
```
Comment (amnonkhen) [inline .gitlab-ci.yml]: the dev file calls the requirements file.
```suggestion
    
```

File: broker/submissions/routes.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    'submissions', __name__, url_prefix='/submissions'
+)
+
+
+@submissions_bp.route('/<submission_uuid>/spreadsheet', methods=['GET'])
+def export_to_spreadsheet(submission_uuid):
+    workbook = ExportToSpreadsheetService(app.ingest_api).export(submission_uuid)
+    timestamp = time.strftime("%Y%m%d-%H%M%S")
+    temp_file = tempfile.NamedTemporaryFile()
+    filename = f'{submission_uuid}_{timestamp}.xlsx'
```
Comment (amnonkhen) [inline broker/submissions/routes.py:21]: What happens fi 2 users try to export the same project? can this cause a problem? Do we need to make this in a temp directory?

**Review by amnonkhen — COMMENTED** [review]
Does this include the "protocol" tests where there is an agreed spreadhseet the works with the import and export?

**Review by ke4 — DISMISSED** [review]

File: broker/service/spreadsheet_upload_service.py (hunk 2 above)
Comment (amnonkhen) [reply broker/service/spreadsheet_upload_service.py:40]: If this is needed during the operation we can store, but we would need to delete the files once the operation is complete. `tempfile` could be used for that.

**Comment by aaclan-ebi** [comment]
> Does this include the "protocol" tests where there is an agreed spreadsheet the works with the import and export

@amnonkhen  There is a task for creating the integration test for this, that should cover that.

File: broker/submissions/routes.py (hunk 5 above)
Comment (aaclan-ebi) [reply broker/submissions/routes.py:21]: it shouldn't be a problem, there is a timestamp (up to seconds) in the filename as well. If that happens the user can just try downloading again. It's already a tempfile object. They get cleaned up automatically.

File: .gitlab-ci.yml (hunk 4 above)
Comment (aaclan-ebi) [reply .gitlab-ci.yml]: Removed.

File: broker/service/spreadsheet_upload_service.py (hunk 2 above)
Comment (aaclan-ebi) [reply broker/service/spreadsheet_upload_service.py:40]: I filed a ticket for that work to remove storing original and updated spreadsheets, we need to check with @clairerye if we do don't want to store the spreadsheets being uploaded by the user
https://github.com/ebi-ait/dcp-ingest-central/issues/395

File: broker/service/spreadsheet_upload_service.py (hunk 1 above)
Comment (aaclan-ebi) [reply broker/service/spreadsheet_upload_service.py:44]: Removed.

File: requirements.txt (hunk 3 above)
Comment (aaclan-ebi) [reply requirements.txt]: Removed.

**Review by ke4 — DISMISSED** [review]

**Review by ke4 — DISMISSED** [review]

**Review by jacobwindsor — DISMISSED** [review]

**Review by ke4 — APPROVED** [review]

**Review by aaclan-ebi — APPROVED** [review]

**Pull request merged**